File: 975-Range Sum of BST.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int res = 0;
    int rangeSumBST1(TreeNode* root, int low, int high) {
        helper(root, low, high);
        return res;
    }
    
    int rangeSumBST(TreeNode *root, int low, int high){
        int res = 0;
        stack<TreeNode> c;
        c.push(*root);
        
        while(!c.empty()){
            TreeNode cur = c.top();
            c.pop();
            if (cur.val >= low && cur.val <= high){
                res += cur.val;
            }
            if(cur.left)
                c.push(*cur.left);
            if (cur.right)
                c.push(*cur.right);
        }
        return res;
    }
    void helper (TreeNode *node, int l, int h){
            if (!node)
                return;
            if (node->val >= l && node->val <= h){
                res += node->val;
            }
            helper(node->left, l, h);
            helper(node->right, l, h);
    }
        
        
};

File: 975-Range Sum of BST.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rangeSumBST(self, root: Optional[TreeNode], low: int, high: int) -> int:
        self.res = 0
        
        def helper(node):
            if not node:
                return
            
            if low <= node.val <= high:
                self.res += node.val
            for child in [node.left, node.right]:
                helper(child)
        
        helper(root)
        return self.res
    
    
    def rangeSumBST(self, root, low, high):
        st = [root]
        res = 0
        while st:
            cur = st.pop()
            if not cur:
                continue
            if low <= cur.val <= high:
                res += cur.val
            
            for child in [cur.left, cur.right]:
                st.append(child)
        
        return res
    
    
    def rangeSumBSTMorris(self, root, l, h):
        res = 0
        
        while root:
            if root.left:
                temp = root.left
                while temp.right:
                    temp = temp.right
                
                temp.right = root
                temp = root.left
                root.left = None
                root = temp
            else:
                if l <= root.val <= h:
                    res += root.val
                root = root.right
        
        return res